crates/wasmparser/src/validator/operators.rs

```diff
@@ -777,6 +876,47 @@ impl OperatorValidator {
                 }
                 self.check_call_indirect(index, table_index, resources)?
             }
+            Operator::CallRef => {
+                self.check_function_references_enabled()?;
+                let rt = self.pop_ref(resources)?;
+                match rt.heap_type {
+                    HeapType::Index(type_index) =>
+                    {
+                        let ft = func_type_at(resources, type_index)?;
+                        for ty in ft.inputs().rev() {
+                            self.pop_operand(Some(ty), resources)?;
+                        }
+                        for ty in ft.outputs() {
+                            self.push_operand(ty, resources)?;
+                        }
+                    },
+                    HeapType::Bot => (),
+                    _ => bail_op_err!("type mismatch: instruction requires function reference type but stack has {}", ty_to_str(ValType::Ref(rt)))
+                }
+            }
+            Operator::ReturnCallRef => {
+                self.check_function_references_enabled()?;
+                if !self.features.tail_call {
+                    return Err(OperatorValidatorError::new(
+                        "tail calls support is not enabled",
+                    ));
+                }
```

Comment:
Could this be extracted to a `check_tail_call_enabled()` helper function in the same manenr as `check_function_references_enabled`?